### PR TITLE
Fix gpexpand: let it disconnect from the database

### PR DESCRIPTION
**Pull request by volkovandr**
The problem is the following: If you try to expand an existing cluster with _gpexpand_ it will connect to the database to analyse the configuration of the cluster and make the changes. But it does not close the session before trying to stop the database. The gpstop will initiate a smart shutdown process, wait 120 seconds and then kill the database with SIGQUIT. After that _pgcontroldata_ returns the state of the cluster "In production" therefore _gpexpand_ thinks it has failed to stop the database and exits.

Another related problem is that when the connection to the database is not closed the script does not return to the command line when it finishes. I don't know the exact reason for this but explicit disconnecting helps.

### CODE REVIEW

File: gpMgmt/bin/gpexpand [hunk 1]
```diff
@@ -705,6 +705,11 @@ class SegmentTemplate:
 
         self.oldSegCount = self.gparray.get_segment_count()
 
+        try:
```
Comment (ascherbaum-pivotal) [inline gpMgmt/bin/gpexpand]: Can this fail? And if yes, isn't it the problem you are trying to fix here?

File: gpMgmt/bin/gppylib/gpcatalog.py [hunk 2]
```diff
@@ -31,7 +31,6 @@ class GPCatalogException(Exception):
     'gp_fault_strategy',
     'gp_san_configuration',
     'gp_segment_configuration',
-    'gp_verification_history',
```
Comment (ascherbaum-pivotal) [inline gpMgmt/bin/gppylib/gpcatalog.py]: What is this line doing, or why is it removed?

Comment (volkovandr) [inline gpMgmt/bin/gppylib/gpcatalog.py]: This is a list called MASTER_ONLY_TABLES. Defined in _gpcatalog.py_ and used in _gpexpand_ only, like this:
`        statements = [ "delete from pg_catalog.%s" % tab for tab in MASTER_ONLY_TABLES ]
`
to delete from those tables.
The table gp_verification_history was excluded from the catalog (see https://github.com/greenplum-db/gpdb/commit/8afc1dd138ec5effcf4cae1799a0d3b9d76aa2a5) but _gpexpand_ still tried to delete from this table and failed with "relation does not exists"

File: gpMgmt/bin/gpexpand (hunk 1 above)
Comment (volkovandr) [inline gpMgmt/bin/gpexpand]: Strictly speaking the connection should be already established at this point and closing it should work fine. The script used to fail at the next line
`GpStop.local('gpexpand _create_template stop gpdb', masterOnly=True, fast=True)`

However i've put it in the try-except block just to be sure that it is not the closing of the connection that makes the script fail.

**Comment by volkovandr** [comment]
Could someone retrigger a build here: https://travis-ci.org/greenplum-db/gpdb/builds/141117029 ?
Looks like it failed to download the repository.

File: gpMgmt/bin/gpexpand (hunk 1 above)
Comment (ascherbaum-pivotal) [inline gpMgmt/bin/gpexpand]: If the connection is established, and you close it and it fails for whatever reason (something hangs, as example) ... you are back at start, right? You still have an open connection, and likely your next operation will fail.

I'm not sure how to handle this correctly, but "pass" looks fishy. Maybe also check if the connection is still open, only close it if it's open and if that fails then bail out with an error message?

File: gpMgmt/bin/gppylib/gpcatalog.py (hunk 2 above)
Comment (ascherbaum-pivotal) [inline gpMgmt/bin/gppylib/gpcatalog.py]: So it's a separate problem, and should be in a separate PR?

**Comment by ascherbaum-pivotal** [comment]
Any code update on the PR will trigger a new build, plus we will test the PR on more systems before committing it.

File: gpMgmt/bin/gppylib/gpcatalog.py (hunk 2 above)
Comment (volkovandr) [inline gpMgmt/bin/gppylib/gpcatalog.py]: I have to admit that you are right.
And i've found another PR https://github.com/greenplum-db/gpdb/pull/895 that also solves the problem of not returning to the command line that was submitted almost at the same time with mine.
Let me then split this one into two, and exclude the code that overlaps with #895

File: gpMgmt/bin/gpexpand (hunk 1 above)
Comment (volkovandr) [inline gpMgmt/bin/gpexpand]: Agree, makes sense. I've changed this in the following commit.

**Comment by volkovandr** [comment]
Since this PR is partially overlapping with #895 i have reverted the file to its original state and made only the changes that are not covered by #895.
Also reverted changes to gpcatalog.py since they are dealing with another problem. Will submit another PR for them.

Now this pull request covers only disconnecting from the database before trying to stop it.
Sorry for confusion.

**Comment by yezhiweicmss** [comment]
hi, @volkovandr 
aha, i met the same problem with you when i tried to "gpexpand -i gpexpand_inputfile -D gpadmin".
the gpexpand restart the GPDB system in utility mode, and made a connection to the system. This causes the problem in SegmentTemplate::_create_template when gpexpand try to GpStop.

your solution works OK.:+1:

**Comment by kaknikhil** [comment]
@volkovandr : we will review and run a few tests and merge it in once everything is good. Thanks for the contribution.

File: gpMgmt/bin/gpexpand [hunk 3]
```diff
@@ -705,6 +705,8 @@ class SegmentTemplate:
 
         self.oldSegCount = self.gparray.get_segment_count()
 
+        self.conn.close()
```
Comment (kaknikhil) [inline gpMgmt/bin/gpexpand]: @volkovandr  as you mentioned before , gpstop -a -m -f does not fully stop the cluster and the state is still "in production" hence gpexpand fails. So your solution to close the connection before gpstop seems reasonable. But why not call self.conn.close() every time we call GpStop.local() ? why be limited to these three instances?

**Comment by kaknikhil** [comment]
@volkovandr i saw in one of your previous commits that you implemented the __del__ function which you then took out because it was taken care of by #895 .  I think having a __del__  which does a "self.conn.close()" is still valuable.

File: gpMgmt/bin/gpexpand (hunk 3 above)
Comment (volkovandr) [inline gpMgmt/bin/gpexpand]: But it is called every time now before GpStop.local()
The only place where it is not so is in 
```
@staticmethod
    def get_gpdb_in_state(state):
```
But there is no _self.conn_ in a staticmethod and also in this function (and in other functions that call this one) the connection to the database is not established.

**Comment by volkovandr** [comment]
The thing was that just having a del function is not enough. It was necessary to explicitly delete the gpexpand instance which I found not nice but I didnt have a better solution. However #895 fixes the problem itself when my del function was not more than just a workaround. That's why I've taken it out.
And now we have three different pull requests that solve three different problems, which I think is right.

**Comment by kaknikhil** [comment]
> The thing was that just having a del function is not enough. It was necessary to explicitly delete the gpexpand instance which I found not nice but I didnt have a better solution.

what i meant was that although #895 fixes the problem , we can still have a __del__ function which closes the  connection at the end. But we don't need to worry about that in this PR.

File: gpMgmt/bin/gpexpand (hunk 3 above)
Comment (kaknikhil) [inline gpMgmt/bin/gpexpand]: yes you are right, i missed the static declaration.

**Comment by kaknikhil** [comment]
Pushed to master https://github.com/greenplum-db/gpdb/commit/51a46a68813084c096605a7ea2090f13f8bbf80a with comment updates. Thanks for your contribution.
